File: typescript/7-seminar-shop/src/ThreeLetterDiscount.ts
import { Seminar } from "./Seminar";
import { THREE_LATER_DISCOUNT_RATE } from "./constants";

export class ThreeLetterDiscount {
  constructor(private seminar: Seminar) {}

  apply(): void {
    this.seminar.netPrice = this.seminar.originalNetPrice - this.discount();
  }

  isGranted(): boolean {
    this.seminar.title;
    return this.seminar.title.length <= 3;
  }

  discount(): number {
    return this.seminar.originalNetPrice * this.discountRate();
  }

  discountRate(): number {
    return !this.isGranted() ? 0 : THREE_LATER_DISCOUNT_RATE / 100;
  }
}

File: python/7-seminar-shop/src/three_letter_discount.py
from seminar import Seminar
from constants import THREE_LETTER_DISCOUNT_RATE


class ThreeLetterDiscount:
    def __init__(self, seminar: Seminar):
        self._seminar = seminar

    def apply(self):
        self._seminar.net_price = self._seminar.original_net_price - self.discount()

    def is_granted(self) -> bool:
        self._seminar.title
        return len(self._seminar.title) <= 3

    def discount(self) -> float:
        return self._seminar.original_net_price * self.discount_rate()

    def discount_rate(self) -> float:
        return 0 if not self.is_granted() else THREE_LETTER_DISCOUNT_RATE / 100
